sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducerAsyncClient.java

```diff
@@ -447,18 +514,65 @@ public String getEventHubName() {
             // Start send span and store updated context
             parentContext.set(tracerProvider.startSpan(finalSharedContext, ProcessKind.SEND));
         }
-
-        return withRetry(getSendLink(batch.getPartitionId())
-            .flatMap(link ->
-                messages.size() == 1
+        if (isIdempotentPartitionPublishing) {
+            PartitionPublishingState publishingState = this.getClientPartitionPublishingState(batch.getPartitionId());
+            return Mono.fromRunnable(() -> {
+                publishingState.getSemaphore().acquireUninterruptibly();
```

Comment:
@srnagar I'm torn on using a semaphore for this. thoughts?

Reply:
The requirement is for a combination of a producer group and a partition, the send should be sequential. No concurrency is allowed because the service will check the producer sequence number assigned by the client. 
I spent long time in searching for a replacement of Semaphore for this use case. Any suggestions?

Reply:
The other option was to consider Reentrant lock but since the thread that acquires the lock may not be the same as the one that releases it, semaphore is the only option here.